docs/proposals/device-crd.md

```diff
@@ -0,0 +1,554 @@
+
+### Use Cases
+
+* Describe device properties.
+  * Users can describe device proeprties and access mechanisms to interact with / control the device.
+* Perform CRUD operations on devices from cloud.
+   * Users can create, update and delete device metadata from the cloud via the CRD APIs exposed by the Kubernetes API server.
+   * Users can control the expected state of a device via the CRD APIs.
+* Report device status from the edge node.
+  * Mapper applications running on the edge can report the device state as well the actual values of the control attributes.
+
+## Design Details
+
+### Device model CRD
+<img src="../images/device-crd/device-model-crd.png">
+
+A `device model` describes the device properties exposed by the device and property visitors to access these properties. A device model is like a reusable template using which many devices can be created and managed.
+
+### Device Model Type Definition
+```go
+// DeviceModelSpec defines the model / template for a device.It is a blueprint which describes the device
+// capabilities and access mechanism via property visitors.
+type DeviceModelSpec struct {
+  // List of device attributes.
+  // Required.
+  DeviceProperties       []DeviceProperty        `json:"properties,omitempty"`
+  // List of property visitors which describe how to access the device attributes.
+  // Required.
+  DevicePropertyVisitors []DevicePropertyVisitor `json:"propertyVisitors,omitempty"`
+}
+
+// DeviceProperty describes an individual device property / attribute like temperature / humidity etc.
+type DeviceProperty struct {
+  // The property name.
+  // Required.
+  Name            string            `json:"name,omitempty"`
+  // The property description.
+  // +optional
+  Description     string            `json:"description,omitempty"`
+  // Additional configuration about the device property like datatype, range, min, max etc.
+  // Optional: If no configuration details are provided, it is assumed that the property is an empty read-only string.
+  // +optional
+  PropertyConfig  map[string]string `json:"propertyConfig,omitempty"`
+}
+
+// DevicePropertyVisitor describes the specifics of accessing a particular device
+// property. Visitors are intended to be consumed by device mappers which connect to devices
+// and collect data / perform actions on the device. A device may support multiple protocols.
+// We can define multiple visitors per protocol for a device. For e.g If a device supports Bluetooth
+// protocol , the visitors would describe the 128-bit UUIDs to read / write charactersitics of a Bluetooth
+// service exposed by the device.
+type DevicePropertyVisitor struct {
+  // The protocol type to connect to the device and access the property.
+  // Required.
+  Protocol     ProtocolType      `json:"protocol,omitempty"`
+  // A list of visitors describing the properties which can be accessed with the protocol.
+  // Required.
+  Visitors     []PropertyVisitor `json:"visitors,omitempty"`
```

Comment:
Define a visitor array in visitor is not a good idea.

Reply:
the top level visitor is per protocol , within each protocol visitor , we have property visitors which can be read using that protocol.

```yaml
    - protocolName: opc-ua    // this is the top level visitor supporting opc-ua
      visitors:
      - name: temperature
        propertyName: Temperature
      - name: pressure
        propertyName: Pressure
      - name: humidity
        propertyName: Humidity
        
    - protocolName: modebus // this is the top level visitor supporting modbus
      visitors:
      - name: humidity
        propertyName: Humidity
```

Reply:
updated APIs don't have this struct now.